internal/sdk/framework_resource_wrapper.go

```diff
@@ -0,0 +1,310 @@
+		parser := resourceids.NewParserFromResourceIdType(id)
+		idVal := ""
+		response.State.GetAttribute(ctx, path.Root("id"), &idVal)
+		parsed, err := parser.Parse(idVal, true)
+		if err != nil {
+			response.Diagnostics.AddError("parsing resource ID: %s", err.Error())
+		}
+
+		segments := id.Segments()
+		numSegments := len(segments)
+		for idx, segment := range segments {
+			if segmentTypeSupported(segment.Type) {
+				name := segmentName(segment, idType, numSegments, idx)
+
+				field, ok := parsed.Parsed[segment.Name]
+				if !ok {
+					response.Diagnostics.AddError("setting resource identity", fmt.Sprintf("field `%s` was not found in the parsed resource ID %s", name, id))
+					return
+				}
+
+				response.Identity.SetAttribute(ctx, path.Root(name), basetypes.NewStringValue(field))
+			}
+		}
+	}
+}
+
+// SetIdentityOnRead sets the identity on the read response based on the resource ID.
+func (r *FrameworkResourceWrapper) SetIdentityOnRead(ctx context.Context, response *resource.ReadResponse) {
+	if id, idType := r.FrameworkWrappedResource.Identity(); id != nil {
+		parser := resourceids.NewParserFromResourceIdType(id)
+		idVal := ""
+		response.State.GetAttribute(ctx, path.Root("id"), &idVal)
+		parsed, err := parser.Parse(idVal, true)
+		if err != nil {
+			response.Diagnostics.AddError("parsing resource ID: %s", err.Error())
+		}
+
+		segments := id.Segments()
+		numSegments := len(segments)
+		for idx, segment := range segments {
+			if segmentTypeSupported(segment.Type) {
+				name := segmentName(segment, idType, numSegments, idx)
+
+				field, ok := parsed.Parsed[segment.Name]
+				if !ok {
+					response.Diagnostics.AddError("setting resource identity", fmt.Sprintf("field `%s` was not found in the parsed resource ID %s", name, id))
+					return
+				}
+
+				response.Identity.SetAttribute(ctx, path.Root(name), basetypes.NewStringValue(field))
+			}
+		}
+	}
+}
+
+// AssertResourceModelType is a helper function to assist in checking the Resource or Data Source model type and
+// provides the compiler with the struct type to be able to access the struct fields correctly.
+// model is the initialised struct, and response is the operation response for which the model is being checked to
+// allow error diagnostics to be written back to the SDK and passed to Terraform.
+func AssertResourceModelType[T any](model any, response any) *T {
+	result, ok := model.(*T)
+	if !ok {
+		switch v := response.(type) {
+		case *resource.CreateResponse:
+			v.Diagnostics.AddError("resource had wrong model type", fmt.Sprintf("got %T", model))
+		case *resource.ReadResponse:
+			v.Diagnostics.AddError("resource had wrong model type", fmt.Sprintf("got %T", model))
+		case *resource.UpdateResponse:
+			v.Diagnostics.AddError("resource had wrong model type, ", fmt.Sprintf("got %T", model))
+		case *resource.DeleteResponse:
+			v.Diagnostics.AddError("resource had wrong model type", fmt.Sprintf("got %T", model))
+
+		case *datasource.ReadResponse:
```

Comment:
Are we supposed to check datasource here?

Reply:
Arguably this func should live in another file as it works for Resources or Data Sources, but I think it's fine here for now. As we collect more such functions, we can look to refactor these into another location?